File: 0268-missing-number/268.py
class Solution:
    def missingNumber(self, nums: List[int]) -> int:
        # Approach 2: XoR Technique -- T.C: O(N), S.C: O(1)
        missing = len(nums)

        for i, num in enumerate(nums):
            missing ^= i ^ num

        return missing

        """
        # Appraoch 1: Math Algebra -- T.C: O(N), S.C: O(1)

        n = len(nums)
        total_sum = (n * (n + 1)) >> 1

        for num in nums:
            total_sum -= num

        return total_sum
        """

File: 0268-missing-number/268.rs
impl Solution {
    pub fn missing_number(nums: Vec<i32>) -> i32 {
        // Approach 1: Xor Technique -- T.C: O(N), S.C: O(1)
        
        let mut missing: i32 = nums.len() as i32;

        for (i, num) in nums.iter().enumerate() {
            missing ^= i as i32 ^ num;
        }
        missing
    }
}
